### ISSUE TITLE
Throwing an error in a server route causes an infinite loop when fetched inside a server only plugin

### ISSUE THREAD

**Issue by damien-hl** [nuxt/nuxt#20909]
### Environment

- Operating System: `Darwin`
- Node Version:     `v18.16.0`
- Nuxt Version:     `3.5.0`
- Nitro Version:    `2.4.1`
- Package Manager:  `pnpm@8.5.0`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`

### Reproduction

https://stackblitz.com/edit/github-v83zpl-v8bzgm

### Describe the bug

When I fetch a server route endpoint `server/routes/****.get.ts` inside a server plugin `plugins/****.server.ts`, **if the server route throw an error, it causes an infinite loop**, calling the the plugin and fetching data from the server route again and again.

<img width="710" alt="infinite_loop" src="https://github.com/nuxt/nuxt/assets/39965922/cf5197f2-fc2a-47e7-8299-dc6f9b8ae3e0">

### Additional context

It does not happen if its an api server route `server/api/****.get.ts` or inside a client plugin `plugins/****.client.ts`.

Catching the error inside the server plugin does not prevent this behaviour :

```ts
try {
  await $fetch('/hello');            
} catch (error) {
  console.error(error);    
}
```

**Comment by danielroe** [comment]
Yes. The issue is that in order to render the _error page_ (being returned to `$fetch('/hello')`, Nuxt will turn to your `error.vue`, and still run all the same plugins... Hence, infinite recursion.

I think we could look to add some kind of detection for this, to improve DX.

**Comment by DengSihan** [comment]
Same issue here, occurred on `useFetch()`

### Describe the bug

When the API throws an error like a `422` or `500` response, `useFetch` will send infinite loop requests. 

![infinite loop requests](https://github.com/nuxt/nuxt/assets/33658414/187aabd0-de84-4b77-9f2c-65c188cb96b9)


### Miniest Reproduction

In `/pages/index.vue`

```ts
<script lang="ts" setup>
const response = await useFetch(`/this-api-will-return-a-422-response`);
</script>
```

**Comment by philippeducasse** [comment]
hi @danielroe , I would like to take over this issue, is it still relevant?

**Comment by stephanelgrg** [comment]
Hello there, 

**Here is what I understand of this issue:**
When a server-only plugin calls $fetch() for a server route and that route throws, Nuxt would still initialize plugins during the server error render. That caused the plugin to run again while rendering the error response, leading to an infinite recursion loop.

**So my idea would be to:**
- Detect server-side error renders in entry.ts
- Skip applyPlugins(nuxt, plugins) when ssrContext.payload.error or ssrContext.error is present

Do you think, this is the way to solve this? 🙏

**Comment by stephanelgrg** [comment]
Hello 👋 

I'm still investigating this issue, and today I found something interesting.
In the providing example: the API is in `playground/server/routes/hello.get.ts` causing a loop like this one:

<img width="1507" height="931" alt="Image" src="https://github.com/user-attachments/assets/c9633566-df71-491a-b577-14c7f4c1d897" />



**But**, following this doc: https://nuxt.com/docs/4.x/directory-structure/server
we can / should declare the endpoint in: `playground/server/api/hello.ts` and it's having a direct impact on this issue:

<img width="1507" height="931" alt="Image" src="https://github.com/user-attachments/assets/e1562b49-c497-4475-919a-d0019a9dde58" />


We can see multiple calls, but no loop anymore 👏 

➡️ Direct fix: use /server/api instead of /server/routes
➡️ I'll keep on investigating this slowly 🐌
